OAuth Signin & Schul-Cloud Integration

jthomaschewski opened the pull request:
Implement login using Schul-Cloud Hydra OAuth2 OpenID Connect Provider.

New Features introduced in this PR:
- "Login mit Schul-Cloud"-Button, Single-Signon with Schul-Cloud Account
- Disable E-Mail/Password fields for Schul-Cloud users
- Add config for disabling password-signin a2nd/or OAuth-signin methods
- Display iFrame from Schul-Cloud with currently logged in user name (email/name is never transferred from Schul-Cloud!)

New env-Configurations:
- `SCHULCLOUD_HYDRA_URL` 
  URL to Hydra/OpenID-Connect Instance of Schul-Cloud. e.g `https://schul-cloud.org/hydra`
- `SCHULCLOUD_CLIENT_ID` / `SCHULCLOUD_CLIENT_SECRET` 
  OAuth2 Client Credentials of Schul-Cloud
- `PW_LOGIN_ENABLED=false/true`
  Enable/Disable email/password login and registration
- `OAUTH_LOGIN_ENABLED=false/true`
  Enable/Disable OAuth2/Schul-Cloud single sign-on

New dependencies:
- `laravel/socialite` (OAuth Signin flow)
- `doctrine/dbal` (Required for change() in migrations)

New Migrations:
- `CreateOAuthIdentitiesTable`
Adds new table `oauth_identities` which maps userIds from OAuth providers to users in `user`-table

- `UpdatePasswordEmailNullableUserTable`
Users who signed up using OAuth don't have a password and not necessary an email address (Schul-Cloud doesn't expose email) => Make fields `password`, `email` nullable

Notes: Updated less-styles to position Schul-Cloud iFrame. Requires rebuild of wepack resources

Tests: None so far. Test environment not configured in this project.

Deployment:

Update image from this PR is available in docker hub: https://hub.docker.com/r/jbbr/portfolio

`jbbr/portfolio:oauth_signin`

jthomaschewski committed:
Implement login via OAuth Provider

Login via Github for now, Hydra provider will be added soon.

New dependencies:
- laravel/socialite (OAuth Signin flow)
- doctrine/dbal (Required for change() in migrations)

jthomaschewski commented:
Wie ist denn bei diesem Projekt das Branching gedacht?

Ich habe die Branches immer mit `development` als Basis erstellt. Ist das so richtig, oder ist `master` die Referenz?
Aktuell scheinen master und development auf dem nahezu selben Stand zu sein. Daher ist ein Rebase im Zweifel schnell gemacht.

aboettger-tuhh commented:
> Wie ist denn bei diesem Projekt das Branching gedacht?
> 
> Ich habe die Branches immer mit `development` als Basis erstellt. Ist das so richtig, oder ist `master` die Referenz?
> Aktuell scheinen master und development auf dem nahezu selben Stand zu sein. Daher ist ein Rebase im Zweifel schnell gemacht.

Das Repository wird aus dem GitLab der TUHH gespiegelt. Dort ist `development` der *Default Branch*. Aus diesem Grund ist es gut, auch hier `development` als Basis zu verwenden.

jthomaschewski committed:
Add Schul-Cloud socialite provider (OAuth2)

new env vars:
SCHULCLOUD_CLIENT_ID
SCHULCLOUD_CLIENT_SECRET
SCHULCLOUD_HYDRA_URL

jthomaschewski committed:
Fix auth issues in SchulCloud Provider

jthomaschewski committed:
Let OAuth login overwrite current login if any

jthomaschewski committed:
Add config for disabling login methods

introduces new env settings:
OAUTH_LOGIN_ENABLED=true
PW_LOGIN_ENABLED=true

jthomaschewski committed:
Make user email optional

remove oauth random email workaround

jthomaschewski committed:
Only allow access to enabled OAuth providers

jthomaschewski committed:
Add support for custom Schul-Cloud iframe for user data

Displays username in iframe above header

jthomaschewski committed:
Small fixes & cleanup

aboettger-tuhh commented:
Sorry für die Verzögerungen. Da ich das System nur administriere, kann ich nicht allein entscheiden. Ich habe die Verantwortlichen ifnormiert.

jthomaschewski commented:
Hallo, kein Problem, ich hab ja grade erst den WIP-Status entfernt :) 

Der nötige OAuth-PR in der Schul-Cloud ist in funktionierende Form noch nicht gemerged.
Daher funktioniert insbesondere die iFrame-Integration aktuell nur mit einer Testinstanz. Wenn kein iFrame übertragen wird, führt das aber auch zu keinem Fehler.

Der OAuth2 Provider für Socialite selber ist aber Hydra/OpenID Conect kompatibel.
Es ist also nicht zu erwarten, dass es hier noch inkompatible Änderungen seitens Schul-Cloud geben wird.

jthomaschewski committed:
Update iframe data on login

make sure possible changes in Schul-Cloud iFrame are saved.

jthomaschewski committed:
OAuth Identities: Add foreign key to users table

jthomaschewski committed:
Update README.md, add OAuth/Schul-Cloud instructions

jthomaschewski committed:
Merge remote-tracking branch 'origin/development' into feature/oauth_signin

konstantin-edtech commented:
Hi @aboettger-tuhh wir vom Projekt Schul-Cloud haben jetzt Euer Tool im Einsatz, es kommt auch sehr gut an & wir würden uns total freuen, wenn Ihr den PR noch merged, damit wir in Zukunft auch von den Updates profitieren können. :-)

aboettger-tuhh commented:
> Hi @aboettger-tuhh wir vom Projekt Schul-Cloud haben jetzt Euer Tool im Einsatz, es kommt auch sehr gut an & wir würden uns total freuen, wenn Ihr den PR noch merged, damit wir in Zukunft auch von den Updates profitieren können. :-)

Ich kann den Wunsch leider nur weitergeben, da ich nur der Administrator bin, der dieses Tool an der TUHH ausrollt. @hklaffke @KnaufB @jlns
